content/chapters/software-stack/lab/README.md

```diff
@@ -0,0 +1,252 @@
+
+Let's build, run and trace system calls for the program:
+
+```Bash
+razvan@yggdrasil:~/.../software-stack/lab/syscall-wrapper$ ls
+main.c  Makefile  syscall.h  syscall.s
+
+razvan@yggdrasil:~/.../software-stack/lab/syscall-wrapper$ make
+gcc -c -o main.o main.c
+nasm -f elf64 -o syscall.o syscall.s
+cc -nostdlib -no-pie -Wl,--entry=main -Wl,--build-id=none  main.o syscall.o   -o main
+
+razvan@yggdrasil:~/.../software-stack/lab/syscall-wrapper$ ls
+main  main.c  main.o  Makefile  syscall.h  syscall.o  syscall.s
+
+razvan@yggdrasil:~/.../software-stack/lab/syscall-wrapper$ ./main
+Hello, world!
+
+razvan@yggdrasil:~/.../software-stack/lab/syscall-wrapper$ strace ./main
+execve("./main", ["./main"], 0x7ffee60fb590 /* 63 vars */) = 0
+write(1, "Hello, world!\n", 14Hello, world!
+)         = 14
+exit(0)                                 = ?
++++ exited with 0 +++
+```
+
+The trace is similar to the previous example, showing the `write` and `exit` system calls.
+
+By creating system call wrappers as C functions, we are now relieved of the burden of writing assembly language code.
+Of course, there has to be an initial implementation of wrapper functions written in assembly language;
+but, after that, we can use C only.
+
+#### Practice
+
+Update the files in the `syscall-wrapper/` folder to make available the `read` system call as a wrapper.
+Make a call to the `read` system call to read data from standard input in a buffer.
+The call `write` to print data from that buffer.
+
+We can see that it's easier to have wrapper calls and write most of the code in C than in assembly language.
+
+## Arena
+
+Go through the practice items below to hone your skills in working with layers of the software stack.
+
+### System Calls
+
+Enter the `basic-syscall/` folder and go through the practice items below.
+If you get stuck, take a sneak peak at the solutions in the `solution/basic-syscall/` folder.
+
+For debugging, use `strace` to trace the system calls from your program and make sure the arguments are set right.
+
+1. Update the `hello.asm` and / or `hello.s` files to pause the execution of the program before the `exit` system call.
+
+   You need to make the `sys_pause` system call, with no arguments.
+   Find its ID [here](https://blog.rchapman.org/posts/Linux_System_Call_Table_for_x86_64/).
+
+1. Update the `hello.asm` and / or `hello.s` files to read a message from standard input and print it to standard output.
+
+   You'll need to define a buffer in the `data` or `bss` section.
+   Use the `read` system call to read data in the buffer.
+   The return value of `read` (placed in the `rax` register) is the number of bytes read.
+   Use that value as the 3rd argument or `write`, i.e. the number of bytes printed.
+
+   Find the ID of the `read` system call [here](https://x64.syscall.sh/).
+   To find out more about its arguments, see [its man page](https://man7.org/linux/man-pages/man2/read.2.html).
+   Standard input descriptor is `0`.
+
```

Comment:
```suggestion
1. Read the content of a file in chunks of 128 bytes and print it to standard output.

   Alongside `read` and `write` system calls, you'll need to use `open` and `close`.
   Pay attention to the return code of each system call as it might indicate encountered errors.
```

Reply:
You don't need to use `open` and `close`. We're using standard file descriptors.